src/finn/custom_op/fpgadataflow/iodma.py

```diff
@@ -0,0 +1,189 @@
+import numpy as np
+from onnx import TensorProto, helper
+from finn.core.datatype import DataType
+from finn.custom_op.fpgadataflow import HLSCustomOp
+
+
+class IODMA(HLSCustomOp):
+    """Class that corresponds to finn-hlslib DMA function(s)."""
+
+    def __init__(self, onnx_node):
+        super().__init__(onnx_node)
+
+    def get_nodeattr_types(self):
+        my_attrs = {
+            "NumChannels": ("i", True, 0),
+            # FINN input datatype
+            "dataType": ("s", True, ""),
+            # DMA-specific parameters
+            "intfWidth": ("i", False, 32),
+            "burstMode": ("s", False, "increment"),
+            "direction": ("s", False, "in"),
+            # shape describing input vecs per execution
+            "numInputVectors": ("ints", False, [1]),
+        }
+        my_attrs.update(super().get_nodeattr_types())
+        return my_attrs
+
+    def get_normal_input_shape(self):
+        vecs = list(self.get_nodeattr("numInputVectors"))
+        num_ch = self.get_nodeattr("NumChannels")
+        ishape = tuple(vecs + [num_ch])
+        return ishape
+
+    def get_normal_output_shape(self):
+        return self.get_normal_input_shape()
+
+    def get_folded_input_shape(self):
+        shape = list(self.get_normal_input_shape())
+        itype_bits = self.get_input_datatype().bitwidth()
+        intfw = self.get_nodeattr("intfWidth")
+        elems_per_word = intfw / itype_bits
+        fold_depth = round(shape[-1] / elems_per_word)
+        shape[-1] = fold_depth
+        shape.append(elems_per_word)
```

Comment:
I don't fully understand the logic here and L42 with the rounding (up? down?) is a bit alarming. what are the DMA alignment restrictions? what implications are there for packing the input/output arrays? do we assume that the innermost (packed) dimension is divisible by the intfWidth?

note that we have e.g. get_instream_width_padded() in the HLSCustomOp base class (returns input stream width padded to nearest 8 bits as required by AXI stream) which may be relevant here:
https://github.com/Xilinx/finn/blob/dev/src/finn/custom_op/fpgadataflow/__init__.py#L568

Reply:
I've documented some of the restrictions. The folded shapes were a bit tricky on the AXI-MM side since there is no guarantee that the AXI-MM interface width is an integer multiple of the element width. With this in mind, I added width conversion, so now the stream side has the same normal shape as the tensor it connects to, and the folded shape is defined by the stream width. The AXI-MM side has no folded shape (it cannot be defined here) but has the same normal shape as the tensor connected to the stream side.